pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_reconcile.go

```diff
@@ -0,0 +1,242 @@
+	logger = logger.WithValues("apiExportWorkspace", logicalcluster.From(export), "apiExportName", export.Name)
+
+	exportedClaims := sets.NewString()
+	for _, claim := range export.Spec.PermissionClaims {
+		exportedClaims.Insert(claim.String())
+	}
+
+	acceptedClaims := sets.NewString()
+	for _, claim := range apiBinding.Spec.PermissionClaims {
+		if claim.State == apisv1alpha1.CLAIM_ACCEPTED {
+			acceptedClaims.Insert(claim.String())
+		}
+	}
+
+	appliedClaims := sets.NewString()
+	for _, claim := range apiBinding.Status.AppliedPermissionClaims {
+		appliedClaims.Insert(claim.String())
+	}
+
+	expectedClaims := exportedClaims.Intersection(acceptedClaims)
+	unexpectedClaims := acceptedClaims.Difference(expectedClaims)
+	needToApply := expectedClaims.Difference(appliedClaims)
+	needToRemove := appliedClaims.Difference(acceptedClaims)
+	allChanges := needToApply.Union(needToRemove)
+
+	logger.V(6).Info("claim set details",
+		"expected", expectedClaims,
+		"unexpected", unexpectedClaims,
+		"toApply", needToApply,
+		"toRemove", needToRemove,
+		"all", allChanges,
+	)
+
+	var allErrors []error
+	applyErrors := sets.NewString()
+	gotPatchError := false
+
+	for _, s := range allChanges.List() {
+
+		claim, err := apisv1alpha1.PermissionClaimFrom(s)
+		if err != nil {
+			allErrors = append(allErrors, fmt.Errorf("error parsing claim string %q: %w", s, err))
+			continue
+		}
+
+		claimLogger := logger.WithValues("claim", s)
+
+		informer, gvr, err := c.getInformerForGroupResource(claim.Group, claim.Resource)
+		if err != nil {
+			allErrors = append(allErrors, fmt.Errorf("error getting informer for group=%q, resource=%q: %w", claim.Group, claim.Resource, err))
+			if acceptedClaims.Has(s) {
+				applyErrors.Insert(s)
+			}
+			continue
+		}
+
+		claimLogger.V(4).Info("listing resources")
+		objs, err := informer.Informer().GetIndexer().ByIndex(indexers.ByLogicalCluster, clusterName.String())
+		if err != nil {
+			allErrors = append(allErrors, fmt.Errorf("error listing group=%q, resource=%q: %w", claim.Group, claim.Resource, err))
+			if acceptedClaims.Has(s) {
+				applyErrors.Insert(s)
+			}
+			continue
+		}
+
+		claimLogger.V(4).Info("got resources", "count", len(objs))
+
+		var claimErrs []error
+		for _, obj := range objs {
+			objectMeta, err := meta.Accessor(obj)
+			if err != nil {
+				claimErrs = append(claimErrs, fmt.Errorf("error getting meta accessor for %T: %w", obj, err))
+				continue
+			}
+
+			patchLogger := claimLogger.WithValues("objectNamespace", objectMeta.GetNamespace(), "objectName", objectMeta.GetName())
+			patchLogger.V(4).Info("patching to get claim labels updated")
+
+			// Empty patch, allowing the admission plugin to update the resource to the correct labels
+			err = c.patchGenericObject(ctx, objectMeta, gvr, clusterName)
+			if err != nil {
+				gotPatchError = true
+				claimErrs = append(claimErrs, fmt.Errorf("error patching %q %s|%s/%s: %w", gvr, clusterName, objectMeta.GetNamespace(), objectMeta.GetName(), err))
```

Comment:
do we cut this down later to limit the list size? We don't want to log 100s or 1000s of errors.

Reply:
yes